viadot/sources/hubspot.py

```diff
@@ -0,0 +1,184 @@
+        - using full url as endpoint.
+    """
+
+    def __init__(self, credentials: dict, *args, **kwargs):
+        """
+        Create an instance of Hubspot.
+
+        Args:
+            credentials (dict): Credentials to Hubspot API.
+        """
+
+        self.credentials = credentials
+        if credentials is not None:
+            try:
+                self.headers = {
+                    "Authorization": f'Bearer {self.credentials["TOKEN"]}',
+                    "Content-Type": "application/json",
+                }
+            except:
+                raise CredentialError("Credentials not found.")
+
+        super().__init__(*args, credentials=self.credentials, **kwargs)
+
+        self.base_url = self.credentials.get("URL")
+
+    def clean_special_characters(self, value: str = None) -> str:
+        """
+        Function that maps special characters from ASCII to HTML5 equivalents.
+
+        Args:
+            value (str, optional): Value for decoding. Defaults to None.
+
+        Returns:
+            str: Value after decoding to HTML5.
+        """
+
+        mapping = {
+            "!": "%21",
+            '"': "%22",
+            "#": "%23",
+            "$": "%24",
+            "%": "%25",
+            "&": "%26",
+            "(": "%28",
+            ")": "%29",
+            "*": "%2A",
+            "+": "%2B",
+            ",": "%2C",
+        }
+
+        for key in mapping.keys():
+            if key in value:
+                value = value.replace(key, mapping[key])
+
+        return value
+
+    def get_properties_url(self, endpoint: str = None) -> str:
+        """
+        Function that generates URL with all properties ( = detailed information about all avaliable columns for defined endpoint (schema)).
+
+        Args:
+            endpoint (str, optional): Hubspot schema from crm objects (eg. contacts, line_items, ...). Defaults to None.
+
+        Returns:
+            str: Url with properties for defined endpoint.
+        """
+
+        url = f"{self.base_url}/properties/v2/{endpoint}/properties"
+        return url
+
+    def get_api_url(
+        self,
+        endpoint: str = None,
+        filters: Dict[str, Any] = None,
+        properties: List[Any] = None,
+    ) -> str:
+        """
+        Function that generates full url for Hubspot API with defined parametrs.
+
+        Args:
+            endpoint (str, optional): Schema or full url. Defaults to None.
+            filters (Dict[str, Any], optional): Filters defined for the API body in specific order. Defaults to None.
+            properties (List[Any], optional): List of user-defined columns to be pulled from the API. Defaults to None.
+
+        Returns:
+            str: Generated url passed to Hubspot API.
+        """
+
+        if self.base_url in endpoint:
+            url = endpoint
+        else:
+            if endpoint.startswith("/"):
+                endpoint = endpoint[1:]
+            if endpoint.startswith("hubdb"):
```

Comment:
It's more to avoid situation when the user defines the endpoint with "/" -> then when merging with base_url we will have double "/".
I removed 1st if, so it won't be confusing. ;)